core/domain/opportunity_services.py

```diff
@@ -386,3 +386,218 @@ def update_opportunities_with_new_topic_name(topic_id, topic_name):
+        opportunity_models.SkillOpportunityModel
+        .get_skill_opportunities(page_size, cursor))
+    opportunities = []
+    for skill_opportunity_model in skill_opportunity_models:
+        skill_opportunity = (
+            get_skill_opportunity_from_model(
+                skill_opportunity_model))
+        opportunities.append(skill_opportunity.to_dict())
+    return opportunities, cursor, more
+
+
+def create_skill_opportunity(skill_opportunity):
+    """Creates a skill opportunity domain object as a SkillOpportunityModel
+    entity in the datastore.
+
+    Args:
+        skill_opportunity: The skill opportunity domain object to be saved in
+            the datastore.
+    """
+    skill_opportunity_model = opportunity_models.SkillOpportunityModel(
+        topic_id=skill_opportunity.topic_id,
+        topic_name=skill_opportunity.topic_name,
+        skill_id=skill_opportunity.skill_id,
+        skill_description=skill_opportunity.skill_description,
+        question_count=skill_opportunity.question_count,
+    )
+    skill_opportunity_model.put()
+
+
+def update_skill_opportunites_skill_description(skill_id, new_description):
+    """Updates the skill_description of SkillOpportunityModel(s) with
+    new_description.
+
+    Args:
+        skill_id: str. The corresponding skill_id of the opportunity.
+        new_description: str. The new skill_description.
+    """
+    skill_opportunity_models = (
+        opportunity_models.SkillOpportunityModel.get_by_skill_id(skill_id))
+    for model in skill_opportunity_models:
+        model.skill_description = new_description
+        model.put()
+
+
+def delete_skill_opportunities(skill_id):
+    """Deletes the SkillOpportunityModel(s) corresponding to the supplied
+    skill_id.
+
+    Args:
+        skill_id: str. The skill_id corresponding to the to-be-deleted
+            SkillOpportunityModel(s)
+    """
+    skill_opportunity_models = (
+        opportunity_models.SkillOpportunityModel.get_by_skill_id(skill_id))
+    opportunity_models.SkillOpportunityModel.delete_multi(
+        skill_opportunity_models)
+
+
+def add_topic_to_skill_opportunity(topic_id, skill_id):
+    """Adds a topic_id and corresponding topic_name to an existing
+    SkillOpportunityModel with the supplied skill_id if available, creating a
+    new SkillOpportunityModel otherwise.
+
+    Args:
+        topic_id: str. The topic_id of the to-be-added topic.
+        skill_id: str. The skill_id of the SkillOpportunityModel.
+    """
+    skill_opportunity_models = (
+        opportunity_models.SkillOpportunityModel.get_by_skill_id(skill_id))
+    topic = topic_fetchers.get_topic_by_id(topic_id)
+    for model in skill_opportunity_models:
+        if model.topic_id == topic_id:
+            return
+        if model.topic_id is None:
+            model.topic_id = topic_id
+            model.topic_name = topic.name
+            model.put()
+            return
+
+    if skill_opportunity_models:
```

Comment:
I'm not sure I follow the logic in this function.

(a) Why are we expecting more than one skill opportunity model?
(b) Why are we creating a new skill opportunity here if skill_opportunity_models is non-empty?

Reply:
Removed as SkillOpportunities no longer include topic information.